src/test/kotlin/graphics/scenery/tests/examples/volumes/ClientApplication.kt

```diff
@@ -0,0 +1,457 @@
+                position = Vector3f(0.0f, 0.0f, 5.0f)
+            }
+            perspectiveCamera(50.0f, 512, 512)
+            scene.addChild(this)
+        }
+
+        //Step 2: Create necessary video-streaming  components
+        val dummyVolume = DummyVolume()
+        with(dummyVolume) {
+            name = "DummyVolume"
+            transferFunction = TransferFunction.ramp(0.1f, 0.5f)
+            scene.addChild(this)
+        }
+
+        val videoPlane = FullscreenObject()
+        with(videoPlane){
+            name = "VRplane"
+            videoPlane.wantsSync = false
+        }
+
+        val bridge = SwingBridgeFrame("1DTransferFunctionEditor")
+        val transferFunctionUI = TransferFunctionEditor(dummyVolume)
+        bridge.addPanel(transferFunctionUI)
+        transferFunctionUI.name = dummyVolume.name
+        val swingUiNode = bridge.uiNode
+        swingUiNode.spatial() {
+            position = Vector3f(2f,0f,0f)
+        }
+
+        val VDIPlane = FullscreenObject()
+        with(VDIPlane){
+            name = "VDIplane"
+            material().textures["diffuse"] = compute.material().textures["OutputViewport"]!!
+        }
+
+        //Step 4: add Empty Node to scene
+        scene.addChild(switch)
+        var count = 0
+        switch.value = "toVR"
+
+        //Step 5: switching code
+        thread {
+            while (true){
+                count++
+                if(count%100000000000 == 0.toLong()){
+                    count = 0
+                    logger.info("$count")
+                }
+
+                // TODO restore
+//                if (transferFunctionUI.switchTo != "")
+//                    switch.value = transferFunctionUI.switchTo
+
+                if (!currentlyVolumeRendering && switch.value.equals("toVR")){
+
+                    logger.info("Volume Rendering")
+
+                    vdiStreaming = false
+                    scene.addChild(videoPlane)
+                    scene.removeChild(compute)
+                    scene.removeChild(VDIPlane)
+
+                    thread {
+                        decodeVideo(videoPlane)
+                    }
+                    currentlyVolumeRendering = true
+                }
+               else if (currentlyVolumeRendering && switch.value.equals("toVDI")){
+
+                   logger.info("VDI streaming")
+
+                    vdiStreaming = true
+                    scene.addChild(VDIPlane)
+                    scene.addChild(compute)
+                    scene.removeChild(videoPlane)
+                    cam.farPlaneDistance = 20.0f
+
+                    if (firstVDIStream){
+                       thread {
+//                           receiveAndUpdateVDI(compute)
+                       }
+                       firstVDIStream = false
+                    }
+                    currentlyVolumeRendering = false
+                }
+            }
+        }
+    }
+
+    private fun decodeVideo( plane: FullscreenObject){
```

Comment:
`decodeVideo` and `drawFrame` contain some more generic functionality, which could actually be moved to `VideoDecoder`. I guess putting a video frame on an object might help others, too.

Reply:
`decodeVideo` has been moved into `VideoDecoder` in 89890cc111248ebd638b73ab10a35197333d5166 where it is called `decodeFrameByFrame`. I think `drawFrame` is very simple and abstract enough already